pennylane/compiler/python_compiler/transforms/mbqc/outline_state_evolution.py

```diff
@@ -0,0 +1,464 @@
+
+                    # extract ops
+                    rewriter.insertion_point = InsertPoint.before(op)
+                    for qb, idx in list(qubit_to_reg_idx.items()):
+                        extract_op = quantum.ExtractOp(current_reg, idx)
+                        rewriter.insert(extract_op)
+                        qb.replace_by_if(
+                            extract_op.qubit, lambda use: use.operation not in insert_ops
+                        )
+                        # update the qubit_to_reg_idx dict
+                        qubit_to_reg_idx[extract_op.qubit] = idx
+                        del qubit_to_reg_idx[qb]
+
+                case _:
+                    # Handle other operations that might has qreg result
+                    # Note that this branch might not be tested so far as adjoint op is not
+                    # tested so far.
+                    if reg := next(
+                        (reg for reg in op.results if isinstance(reg.type, quantum.QuregType)), None
+                    ):
+                        current_reg = reg
+
+    def _create_state_evolution_function(self, rewriter: pattern_rewriter.PatternRewriter):
+        """Create a new func.func for the state evolution region using clone approach."""
+
+        alloc_op, terminal_boundary_op = self._find_evolution_range()
+
+        # collect operation from alloc_op to terminal_boundary_op
+        ops_to_clone = self._collect_operations_in_range(alloc_op, terminal_boundary_op)
+
+        # collect required inputs for the state evolution funcOp
+        required_inputs = self._collect_required_inputs_for_state_evolution_func(ops_to_clone)
+
+        # colloect required outputs for the state evolution funcOp
+        required_outputs = self._collect_required_outputs(ops_to_clone, terminal_boundary_op)
+
+        register_inputs = []
+        other_inputs = []
+        for val in required_inputs:
+            if isinstance(val.type, quantum.QuregType):
+                register_inputs.append(val)
+            else:
+                other_inputs.append(val)
+
+        register_outputs = []
+        other_outputs = []
+        for val in required_outputs:
+            if isinstance(val.type, quantum.QuregType):
+                register_outputs.append(val)
+            else:
+                other_outputs.append(val)
+
+        ordered_inputs = register_inputs + other_inputs
+        ordered_outputs = register_outputs + other_outputs
+
+        input_types = [val.type for val in ordered_inputs]
+        output_types = [val.type for val in ordered_outputs]
+        fun_type = builtin.FunctionType.from_lists(input_types, output_types)
+
+        state_evolution_func = func.FuncOp(
+            self.original_func_op.sym_name.data + ".state_evolution", fun_type, visibility="private"
+        )
+        rewriter.insert_op(state_evolution_func, InsertPoint.at_end(self.module.body.block))
+
```

Comment:
Would you mind to add your thoughts on the `value_mapper` here @rniczh? I failed to find out useful information on it.